### PR TITLE
[WebAssembly] Add path to PIC mode for wasm tables

### LINKED ISSUE #65191: clang "cannot select" when using wasm table builtins and -fPIC
## Example:
Source file:
```c
static __externref_t table[0];

__externref_t test_builtin_wasm_table_get() {
  return __builtin_wasm_table_get(table, 0);
}
```
Compiling as follows works:
```
clang -c -target wasm32 -mreference-types builtins-table.c
```
but if I add `-fPIC`, it says:
```
fatal error: error in backend: Cannot select: intrinsic %llvm.wasm.table.get.externref
```

## Bad/good IR

The IR looks as follows:
```llvm
@table = internal addrspace(1) global [0 x ptr addrspace(10)] zeroinitializer, align 1

; Function Attrs: noinline nounwind optnone
define hidden ptr addrspace(10) @test_builtin_wasm_table_get() #0 {
entry:
  %0 = call ptr addrspace(10) @llvm.wasm.table.get.externref(ptr addrspace(1) @table, i32 0)
  ret ptr addrspace(10) %0
}

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(read)
declare ptr addrspace(10) @llvm.wasm.table.get.externref(ptr addrspace(1), i32) #1
```
If I change `@table = internal addrspace(1)` to `@table = local_unnamed_addr addrspace(1)` it seems to fix the problem.

@pmatos

### PR DESCRIPTION
**Pull request by pmatos**
Currently tables cannot be shared between compilation units, therefore
no special treatment is needed for tables.

Fixes #65191

### CODE REVIEW

File: llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp [hunk 1]
```diff
@@ -1724,8 +1724,14 @@ SDValue WebAssemblyTargetLowering::LowerGlobalAddress(SDValue Op,
     fail(DL, DAG, "Invalid address space for WebAssembly target");
 
   unsigned OperandFlags = 0;
-  if (isPositionIndependent()) {
-    const GlobalValue *GV = GA->getGlobal();
+  const GlobalValue *GV = GA->getGlobal();
+  // is PIC but not a WebAssembly table.
+  // Since WebAssembly tables cannot yet be shared accross modules, we don't need special
+  // treatment for tables in PIC mode.
```
Comment (sbc100) [inline llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:1730]: Are you saying that shouldAssumeDSOLocal() is always true for table symbols?


Why can't tables be imported/exported like any other symbol?

File: llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp [hunk 2]
```diff
@@ -1724,8 +1724,14 @@ SDValue WebAssemblyTargetLowering::LowerGlobalAddress(SDValue Op,
...
+  if (isPositionIndependent() && 
+      !(GV->getValueType()->isArrayTy() && 
+        WebAssembly::isWebAssemblyReferenceType(GV->getValueType()->getArrayElementType()))) {
```
Comment (sbc100) [inline llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:1733]: How about writing a little helper to make this obvious, then we wouldn't need the comment: `isTableType(GV->getValueType())` ?

File: llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp (hunk 1 above)
Comment (pmatos) [reply llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:1730]: As it stands tables need to be always declared with static. Symbols cannot be exported. I will get around to implementing it at some point to fix https://github.com/llvm/llvm-project/issues/64532.

File: llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp (hunk 2 above)
Comment (pmatos) [reply llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:1733]: That makes sense - will do that.

File: llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp [hunk 3]
```diff
@@ -1724,8 +1724,12 @@ SDValue WebAssemblyTargetLowering::LowerGlobalAddress(SDValue Op,
     fail(DL, DAG, "Invalid address space for WebAssembly target");
 
   unsigned OperandFlags = 0;
-  if (isPositionIndependent()) {
-    const GlobalValue *GV = GA->getGlobal();
+  const GlobalValue *GV = GA->getGlobal();
+  // is PIC but not a WebAssembly table.
```
Comment (sbc100) [inline llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:1728]: I don't think this line is needed anymore

File: llvm/test/MC/WebAssembly/reloc-pic.s [hunk 4]
```diff
@@ -1,5 +1,5 @@
-# RUN: llvm-mc -triple=wasm32-unknown-unknown -filetype=obj < %s | obj2yaml | FileCheck %s
-# RUN: llvm-mc -triple=wasm32-unknown-unknown -mattr=+reference-types -filetype=obj < %s | obj2yaml | FileCheck --check-prefix=REF %s
+# RUN: sed -e '/^REF-/d' %s | llvm-mc -triple=wasm32-unknown-unknown -filetype=obj | obj2yaml | FileCheck %s
+# RUN: sed -e 's/^REF-//g' %s | llvm-mc -triple=wasm32-unknown-unknown -mattr=+reference-types -filetype=obj | obj2yaml | FileCheck --check-prefix=REF %s
```
Comment (sbc100) [inline llvm/test/MC/WebAssembly/reloc-pic.s:2]: Is this something is OK to do in lit test?  I've not seen this sed pattern before.

**Review by sbc100 — COMMENTED** [review]
lgtm % a couple more comments

File: llvm/test/MC/WebAssembly/reloc-pic.s (hunk 4 above)
Comment (pmatos) [reply llvm/test/MC/WebAssembly/reloc-pic.s:2]: We have at least test in WebAssembly (tls-local-exec.ll) using sed. There are many in RISCV which uses sed to replace types depending on current running arch, etc. We need this here to ensure that without reference types we don't add a table and if we do have reference types, the result is correct. Alternatively we need another test file. I am happy to do that if you're not keen on sed but sed is very much used together with lit tests in general.

File: llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp (hunk 1 above)
Comment (pmatos) [reply llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp]: I will wait for a further thumbs up from you if you're happy with introducing this sed pattern into the tests.

File: llvm/test/MC/WebAssembly/reloc-pic.s (hunk 4 above)
Comment (pmatos) [reply llvm/test/MC/WebAssembly/reloc-pic.s:2]: I will wait for a further thumbs up from you if you're happy with introducing this sed pattern into the tests.

**Review by sbc100 — APPROVED** [review]

**Pull request merged**